pkg/sql/copy/copy_test: TestDataDriven failed

pkg/sql/copy/copy_test.TestDataDriven [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9632591?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9632591?buildTab=artifacts#/) on release-23.1 @ [6cce6c746150307a9ecf3b529dfd633a6985c110](https://github.com/cockroachdb/cockroach/commits/6cce6c746150307a9ecf3b529dfd633a6985c110):


```
        ----
        CPut /Table/<>/1/-1/0 -> /TUPLE/
        CPut /Table/<>/1/-1/1/1 -> /INT/-2
    copy_test.go:115: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/1985/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/sql/copy/copy_test_/copy_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/copy/testdata/copy_from:575:
        copy-from-kvtrace [0 args]
        COPY tpartial2 FROM STDIN
        0	0
        ----
        CPut /Table/<>/1/0/0 -> /TUPLE/
        CPut /Table/<>/1/0/1/1 -> /INT/0
    copy_test.go:115: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/1985/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/sql/copy/copy_test_/copy_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/copy/testdata/copy_from:582:
        copy-from-kvtrace [0 args]
        COPY tpartial2 FROM STDIN
        1	2
        ----
        CPut /Table/<>/1/1/0 -> /TUPLE/
        CPut /Table/<>/1/1/1/1 -> /INT/2
        InitPut /Table/<>/2/2/1/0 -> /BYTES/
    copy_test.go:115: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/1985/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/sql/copy/copy_test_/copy_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/copy/testdata/copy_from:590:
        exec-ddl [0 args]
        CREATE TYPE testenum AS ENUM('cat','dog','bear');
        CREATE TABLE tenum (i INT PRIMARY KEY,c1 testenum)
        ----
    copy_test.go:115: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/1985/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/sql/copy/copy_test_/copy_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/copy/testdata/copy_from:595:
        copy-from [0 args]
        COPY tenum FROM STDIN
        0	cat
        1	dog
        2	bear
        ----
        3
    copy_test.go:115: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/1985/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/sql/copy/copy_test_/copy_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/copy/testdata/copy_from:603:
         COPY tenum FROM STDIN
        0	cat
        1	dog
        2	bear
        expected:
        CPut /Table/<>/1/0/0 -> /TUPLE/2:2:Bytes/@
        CPut /Table/<>/1/1/0 -> /TUPLE/2:2:Bytes/0x80
        CPut /Table/<>/1/2/0 -> /TUPLE/2:2:Bytes/0xc0
        
        found:
        CPut /Table/<>/1/0/0 -> /TUPLE/2:2:Bytes/@
        CPut /Table/<>/1/1/0 -> /TUPLE/2:2:Bytes/0x80
    --- FAIL: TestDataDriven/copy_from#07 (1.75s)
```
<p>Parameters: <code>TAGS=bazel,gss</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-queries
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestDataDriven.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-27057

--- a/pkg/sql/copy/copy_test.go
+++ b/pkg/sql/copy/copy_test.go
@@ -143,6 +143,7 @@ func TestDataDriven(t *testing.T) {
 								require.Error(t, err, "copy-from-error didn't return and error!")
 								return err.Error()
 							case "copy-from-kvtrace":
+								require.NoError(t, err, "%s\n%s\n", d.Cmd, d.Input)
 								rows, err := conn.Query(ctx,
 									`SELECT
 						  regexp_replace(message, '/Table/[0-9]*/', '/Table/<>/')

--- a/pkg/sql/copy/testdata/copy_from
+++ b/pkg/sql/copy/testdata/copy_from
@@ -602,13 +602,13 @@ COPY tenum FROM STDIN
 
 copy-from-kvtrace
 COPY tenum FROM STDIN
-0	cat
-1	dog
-2	bear
+3	cat
+4	dog
+5	bear
 ----
-CPut /Table/<>/1/0/0 -> /TUPLE/2:2:Bytes/@
-CPut /Table/<>/1/1/0 -> /TUPLE/2:2:Bytes/0x80
-CPut /Table/<>/1/2/0 -> /TUPLE/2:2:Bytes/0xc0
+CPut /Table/<>/1/3/0 -> /TUPLE/2:2:Bytes/@
+CPut /Table/<>/1/4/0 -> /TUPLE/2:2:Bytes/0x80
+CPut /Table/<>/1/5/0 -> /TUPLE/2:2:Bytes/0xc0
 
 exec-ddl
 CREATE TABLE tenum2 (i INT PRIMARY KEY, c1 testenum, INDEX(c1))
@@ -622,10 +622,10 @@ COPY tenum2 FROM STDIN
 
 copy-from-kvtrace
 COPY tenum2 FROM STDIN
-0	cat
+1	cat
 ----
-CPut /Table/<>/1/0/0 -> /TUPLE/2:2:Bytes/@
-InitPut /Table/<>/2/"@"/0/0 -> /BYTES/
+CPut /Table/<>/1/1/0 -> /TUPLE/2:2:Bytes/@
+InitPut /Table/<>/2/"@"/1/0 -> /BYTES/
 
 exec-ddl
 CREATE TABLE tenum3 (i INT PRIMARY KEY, c1 testenum, UNIQUE INDEX(c1))
@@ -639,10 +639,10 @@ COPY tenum3 FROM STDIN
 
 copy-from-kvtrace
 COPY tenum3 FROM STDIN
-0	cat
+1	dog
 ----
-CPut /Table/<>/1/0/0 -> /TUPLE/2:2:Bytes/@
-InitPut /Table/<>/2/"@"/0 -> /BYTES/0x88
+CPut /Table/<>/1/1/0 -> /TUPLE/2:2:Bytes/0x80
+InitPut /Table/<>/2/"\x80"/0 -> /BYTES/0x89
 
 exec-ddl
 CREATE TYPE comp AS (a INT, b INT);
@@ -657,6 +657,6 @@ COPY tcomp FROM STDIN
 
 copy-from-kvtrace
 COPY tcomp FROM STDIN
-0	(1, 2)
+1	(1, 2)
 ----
-CPut /Table/<>/1/0/0 -> /TUPLE/
+CPut /Table/<>/1/1/0 -> /TUPLE/

--- a/pkg/sql/row/putter.go
+++ b/pkg/sql/row/putter.go
@@ -238,7 +238,13 @@ type kvSparseSliceBulkSource[T kv.GValue] struct {
 var _ kv.BulkSource[[]byte] = &kvSparseSliceBulkSource[[]byte]{}
 
 func (k *kvSparseSliceBulkSource[T]) Len() int {
-	return len(k.keys)
+	cnt := 0
+	for _, k := range k.keys {
+		if len(k) > 0 {
+			cnt++
+		}
+	}
+	return cnt
 }
 
 type sliceIterator[T kv.GValue] struct {
@@ -253,9 +259,13 @@ func (k *kvSparseSliceBulkSource[T]) Iter() kv.BulkSourceIterator[T] {
 }
 
 func (s *sliceIterator[T]) Next() (roachpb.Key, T) {
-	k, v := s.s.keys[s.cursor], s.s.values[s.cursor]
-	s.cursor++
-	return k, v
+	for {
+		k, v := s.s.keys[s.cursor], s.s.values[s.cursor]
+		s.cursor++
+		if len(k) > 0 {
+			return k, v
+		}
+	}
 }
 
 // KVBatchAdapter implements Putter interface and adapts it to kv.Batch API.

